LibraBFT/Impl/Properties/VotesOnceDirect.agda

```diff
@@ -104,65 +104,66 @@ module LibraBFT.Impl.Properties.VotesOnceDirect where
   -- will involve much of the same complexity, and perhaps be more difficult overall.  Maybe
   -- something like the following could be proved inductively, and then used together with
   -- injectivity properties as mentioned above to determine that the two pids are the same.
+  peerCanSignEp≡ : ∀ {pid v v' pk s'}
+                   → PeerCanSignForPK s' v pid pk
+                   → v ^∙ vEpoch ≡ v' ^∙ vEpoch
+                   → PeerCanSignForPK s' v' pid pk
+  peerCanSignEp≡ (mkPCS4PK 𝓔₁ 𝓔id≡₁ 𝓔inSys₁ mbr₁ nid≡₁ pk≡₁) refl
+    = (mkPCS4PK 𝓔₁ 𝓔id≡₁ 𝓔inSys₁ mbr₁ nid≡₁ pk≡₁)
 
-  postulate
-   MsgWithSig⇒ValidSenderInitialised :
-     ∀ {st pk v}
+  MsgWithSig⇒ValidSenderInitialised :
+     ∀ {st v pk}
      → ReachableSystemState st
-     → Meta-Honest-PK pk
+     → Meta-Honest-PK pk → (sig : WithVerSig pk v)
      → MsgWithSig∈ pk (₋vSignature v) (msgPool st)
      → ∃[ pid ] ( initialised st pid ≡ initd
                 × PeerCanSignForPK st v pid pk )
+  MsgWithSig⇒ValidSenderInitialised {st} {v} (step-s r step@(step-peer (step-honest {pid} stP))) pkH sig msv
+     with newMsg⊎msgSentB4 r stP pkH (msgSigned msv) (msg⊆ msv) (msg∈pool msv)
+  ...| inj₁ (m∈outs , pcsN , newV)
+     with stP
+  ...| step-msg _ initP
+      with sameHonestSig⇒sameVoteData pkH (msgSigned msv) sig (msgSameSig msv)
+  ...| inj₁ hb   = ⊥-elim (PerState.meta-sha256-cr st (step-s r step) hb)
+  ...| inj₂ refl = pid , peersRemainInitialized step initP , peerCanSignEp≡ pcsN refl
+  MsgWithSig⇒ValidSenderInitialised {st} {v} (step-s r step@(step-peer (step-honest stP))) pkH sig msv
+     | inj₂ msb4 rewrite msgSameSig msv
+     with MsgWithSig⇒ValidSenderInitialised {v = v} r pkH sig msb4
+  ...| pid , initP , pcsPre = pid ,
+                              peersRemainInitialized step initP ,
+                              PeerCanSignForPK-stable r step pcsPre
+  MsgWithSig⇒ValidSenderInitialised {st} {v} (step-s r step@(step-peer cheat@(step-cheat x))) pkH sig msv
+     with ¬cheatForgeNew cheat refl unit pkH msv
+  ...| msb4
+     with MsgWithSig⇒ValidSenderInitialised {v = v} r pkH sig msb4
+  ...| pid , initP , pcsPre = pid ,
+                              peersRemainInitialized step initP ,
+                              PeerCanSignForPK-stable r step pcsPre
 
 
-  peerCanSign-Msb4 : ∀ {pid v s' outs pk}{st : SystemState}
-                    → ReachableSystemState st
-                    → (stP : StepPeer st pid s' outs)
-                    → PeerCanSignForPK (StepPeer-post stP) v pid pk
+  peerCanSign-Msb4 : ∀ {pid v pk}{pre post : SystemState}
```

Comment:
This property holds and is easy to prove *for now*, given that we have no epoch changes.  However, it will *not* hold when we add epoch changes, because `EpochConfig∈Sys` will have (an) additional constructor(s) that can be established by `stP` (e.g., if the step commits an epoch-changing transaction, resulting in a new `EpochConfig`), in which case it is *not* true that it held in the prestate.  It is used extensively throughout this file, which makes me think the battle against `unwind` is still not over (not to say that it can't be won: I still think you're right that it can, but I think this issue hides some significant challenges).

Reply:
Ooops, sorry, this comment is wrong!  I had overlooked the `MsgWithSig∈ pk (ver-signature sig) (msgPool st)` argument, which should enable this property to be proved (though not quite as easily :smile:) even after we model epoch changes.